File: Artificial Intelligence/Everything Else/hangman.py
import random

WORDS = ["python", "game", "hangman", "test", "winning"]
HANGMAN_STAGES = [
    "  +---+\n  |   |\n      |\n      |\n      |\n      |\n=========",
    "  +---+\n  |   |\n  O   |\n      |\n      |\n      |\n=========",
    "  +---+\n  |   |\n  O   |\n  |   |\n      |\n      |\n=========",
    "  +---+\n  |   |\n  O   |\n /|   |\n      |\n      |\n=========",
    "  +---+\n  |   |\n  O   |\n /|\\  |\n      |\n      |\n=========",
    "  +---+\n  |   |\n  O   |\n /|\\  |\n /    |\n      |\n=========",
    "  +---+\n  |   |\n  O   |\n /|\\  |\n / \\  |\n      |\n=========",
]


def choose_word():
    return random.choice(WORDS)


def display_word(word, guessed):
    return " ".join([letter if letter in guessed else "_" for letter in word])


def play():
    word = choose_word()
    guessed_letters = set()
    attempts = len(HANGMAN_STAGES) - 1
    while attempts > 0:
        print(HANGMAN_STAGES[len(HANGMAN_STAGES) - 1 - attempts])
        print("Word:", display_word(word, guessed_letters))
        print("Guessed Letters:", " ".join(sorted(guessed_letters)))
        guess = input("Enter a letter: ").lower()
        if not guess.isalpha() or len(guess) != 1:
            print("Invalid input. Enter a single letter.")
            continue
        if guess in guessed_letters:
            print("Already guessed.")
            continue
        guessed_letters.add(guess)
        if guess not in word:
            attempts -= 1
            print("Wrong guess!")
        if all(letter in guessed_letters for letter in word):
            print("\nYou Win! Word was:", word)
            return

    print(HANGMAN_STAGES[-1])
    print("\nYou Lost! Word was:", word)


if __name__ == "__main__":
    play()

File: Artificial Intelligence/Everything Else/hangman.pl
% Start the game
start :-
    Word = [t,e,s,t],
    length(Word, Len),
    create_hidden(Len, Hidden),
    play(Word, Hidden, 6).

% Create hidden word
create_hidden(0, []).
create_hidden(N, ['_'|T]) :-
    N > 0,
    N1 is N - 1,
    create_hidden(N1, T).

% Game over case
play(_, _, 0) :-
    write('Game Over! You ran out of attempts.'), nl.

% Main game loop
play(Word, Hidden, Attempts) :-
    Attempts > 0,
    write('Word: '), write(Hidden), nl,
    write('Attempts left: '), write(Attempts), nl,
    write('Enter a letter: '),
    read(Guess),
    update_word(Word, Hidden, Guess, NewHidden),
    (
        Hidden \= NewHidden ->
        write('Correct guess!'), nl,
        NewAttempts = Attempts
        ;
        write('Wrong guess!'), nl,
        NewAttempts is Attempts - 1
    ),
    (
        NewHidden = Word ->
        write('Congratulations! You guessed the word.'), nl
        ;
        play(Word, NewHidden, NewAttempts)
    ).

% Update word
update_word([], [], _, []).
update_word([H|T], [_|T1], Guess, [H|T2]) :-
    H = Guess,
    update_word(T, T1, Guess, T2).
update_word([H|T], [H1|T1], Guess, [H1|T2]) :-
    H \= Guess,
    update_word(T, T1, Guess, T2). 